src/core/fetchers/cdn_prioritizer.ts

```diff
@@ -0,0 +1,410 @@
+      const steeringManifestUpdateCanceller = new TaskCanceller();
+      this._steeringManifestUpdateCanceller = steeringManifestUpdateCanceller;
+    }
+    const canceller : TaskCanceller = this._steeringManifestUpdateCanceller;
+    steeringManifestFetcher.fetch(contentSteering.url,
+                                  (err : IPlayerError) => this.trigger("warnings", [err]),
+                                  canceller.signal)
+      .then((parse) => {
+        const parsed = parse((errs) => this.trigger("warnings", errs));
+        const prevSteeringManifest = this._lastSteeringManifest;
+        this._lastSteeringManifest = parsed;
+        if (parsed.lifetime > 0) {
+          const timeout = window.setTimeout(() => {
+            canceller.signal.deregister(onTimeoutEnd);
+            this._autoRefreshSteeringManifest(steeringManifestFetcher, contentSteering);
+          }, parsed.lifetime * 1000);
+          const onTimeoutEnd = () => {
+            clearTimeout(timeout);
+          };
+          canceller.signal.register(onTimeoutEnd);
+        }
+        if (this._readyState.getValue() === "not-ready") {
+          this._readyState.setValue("ready");
+        }
+        if (canceller.isUsed) {
+          return;
+        }
+        if (prevSteeringManifest === null ||
+            prevSteeringManifest.priorities.length !== parsed.priorities.length ||
+            prevSteeringManifest.priorities
+              .some((val, idx) => val !== parsed.priorities[idx]))
+        {
+          this.trigger("priorityChange", null);
+        }
+      })
+      .catch((err) => {
+        if (err instanceof CancellationError) {
+          return;
+        }
+        const formattedError = formatError(err, {
+          defaultCode: "NONE",
+          defaultReason: "Unknown error when fetching and parsing the steering Manifest",
+        });
+        this.trigger("warnings", [formattedError]);
+      });
+  }
+
+  /**
+   * @param {number} index
+   */
+  private _removeIndexFromDowngradeList(index : number) : void {
+    this._downgradedCdnList.metadata.splice(index, 1);
+    const oldTimeout = this._downgradedCdnList.timeouts.splice(index, 1);
+    clearTimeout(oldTimeout[0]);
+  }
+}
+
+export interface ICdnPrioritizerEvents {
+  warnings : IPlayerError[];
+  priorityChange : null;
+}
+
+/**
+ * Find the index of the given CDN metadata in a CDN metadata array.
+ * Returns `-1` if not found.
+ * @param {Array.<Object>} arr
+ * @param {Object} elt
+ * @returns {number}
+ */
+function indexOfMetadata(
+  arr : ICdnMetadata[],
+  elt : ICdnMetadata
+) : number {
+  if (arr.length === 0) {
+    return -1;
+  }
+  if (elt.id !== undefined) {
+    for (let i = 0; i < arr.length; i++) {
+      const m = arr[i];
+      if (m.id === elt.id) {
+        return i;
+      }
+    }
+  } else {
+    for (let i = 0; i < arr.length; i++) {
+      const m = arr[i];
+      if (m.baseUrl === elt.baseUrl) {
+        return i;
+      }
+    }
+  }
```

Comment:
```typescript
  if (elt.id !== undefined) {
    return arr.findIndex((m) => m.id === elt.id)
  } else {
    return arr.findIndex((m) => m.baseUrl === m.baseUrl)
  }
```

Reply:
Yes it looks like you're right! Definitely more clear that way

Reply:
Done